runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPortablePipelineTranslator.java

```diff
@@ -201,11 +372,171 @@ public void translateImpulse(
+    // Enforce tuple tag sorting by union tag index.
+    Map<String, Coder<WindowedValue<?>>> outputCoders = Maps.newHashMap();
+    for (String localOutputName : new TreeMap<>(outputMap.inverse()).values()) {
+      String collectionId = outputs.get(localOutputName);
+      RunnerApi.PCollection coll = components.getPcollectionsOrThrow(collectionId);
+      RunnerApi.Coder coderProto = components.getCodersOrThrow(coll.getCoderId());
+      Coder<WindowedValue<?>> windowCoder;
+      try {
+        Coder elementCoder = CoderTranslation.fromProto(coderProto, rehydratedComponents);
+        RunnerApi.WindowingStrategy windowProto = components.getWindowingStrategiesOrThrow(
+                coll.getWindowingStrategyId());
+        WindowingStrategy<Object, BoundedWindow> windowingStrategy =
+                (WindowingStrategy<Object, BoundedWindow>)
+                        WindowingStrategyTranslation.fromProto(windowProto, rehydratedComponents);
+        windowCoder = WindowedValue.getFullCoder(elementCoder,
+                windowingStrategy.getWindowFn().windowCoder());
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+      outputCoders.put(localOutputName, windowCoder);
+      unionCoders.add(windowCoder);
+    }
+    //UnionCoder unionCoder = UnionCoder.of(unionCoders);
+    //TypeInformation<RawUnionValue> typeInformation =
+    //        new CoderTypeInformation<>(unionCoder);
+
+    RunnerApi.ExecutableStagePayload stagePayload;
+    try {
+      stagePayload = RunnerApi.ExecutableStagePayload.parseFrom(transform.getSpec().getPayload());
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+
+    String inputPCollectionId =
+            Iterables.getOnlyElement(transform.getInputsMap().values());
+
+    // TODO: is this still relevant?
+    // we assume that the transformation does not change the windowing strategy.
+    RunnerApi.WindowingStrategy windowingStrategyProto =
+            pipeline.getComponents().getWindowingStrategiesOrThrow(
+                    pipeline.getComponents().getPcollectionsOrThrow(
+                            inputPCollectionId).getWindowingStrategyId());
+    final WindowingStrategy<Object, BoundedWindow> windowingStrategy;
+    try {
+      windowingStrategy =
+              (WindowingStrategy<Object, BoundedWindow>) WindowingStrategyTranslation.fromProto(
+                      windowingStrategyProto,
+                      rehydratedComponents);
+    } catch (InvalidProtocolBufferException e) {
+      throw new IllegalStateException(
+              String.format("Unable to hydrate ExecutableStage windowing strategy %s.",
+                      windowingStrategyProto),
+              e);
+    }
+
+    Map<TupleTag<?>, OutputTag<WindowedValue<?>>> tagsToOutputTags = Maps.newHashMap();
+    Map<TupleTag<?>, Coder<WindowedValue<?>>> tagsToCoders = Maps.newHashMap();
+    // TODO: does it matter which output we designate as "main"
```

Comment:
As I understand it, portable pipelines do not have a notion of "main" outputs. This is a quirk of using the non-portable Java support code.

Reply:
Yep, and I'm happy to refactor most of this once we get something running.